metals/src/main/scala/scala/meta/internal/metals/MetalsLanguageServer.scala

```diff
@@ -0,0 +1,746 @@
+            languageClient.showMessage(
+              messages.ImportProjectPartiallyFailed
+            )
+            // Connect nevertheless, many build import failures are caused
+            // by resolution errors in one weird module while other modules
+            // exported successfully.
+            quickConnectToBuildServer()
+          } else {
+            languageClient.showMessage(messages.ImportProjectFailed)
+            Future.successful(BuildChange.Failed)
+          }
+        } else {
+          Future.successful(BuildChange.None)
+        }
+      }
+    } yield change
+
+  private def quickConnectToBuildServer(): Future[BuildChange] = {
+    if (!buildTools.isBloop) {
+      Future.successful(BuildChange.None)
+    } else {
+      val importingBuild = for {
+        _ <- buildServer match {
+          case Some(old) => old.shutdown()
+          case None => Future.successful(())
+        }
+        build <- timed("connected to build server")(bloopServers.newServer())
+        _ = {
+          cancelables.add(build)
+          buildServer = Some(build)
+        }
+        _ <- build.initialize()
+        _ <- installWorkspaceBuildTargets(build)
+      } yield ()
+
+      for {
+        _ <- statusBar.trackFuture("Importing build", importingBuild)
+        _ = statusBar.addMessage(s"${icons.rocket}Imported build!")
+        _ <- compileSourceFiles(buffers.open.toSeq)
+      } yield BuildChange.Reconnected
+    }
+  }.recover {
+    case NonFatal(e) =>
+      val message =
+        "Failed to connect with build server, no functionality will work."
+      val details = " See logs for more details."
+      buildServer.foreach(_.shutdown())
+      buildServer = None
+      scribe.error(message, e)
+      languageClient.showMessage(
+        new MessageParams(MessageType.Error, message + details)
+      )
+      BuildChange.Failed
+  }
+
+  /**
+   * Visit every file and directory in the workspace and register
+   * toplevel definitions for scala source files.
+   */
+  private def indexSourceDirectory(
+      sourceDirectory: AbsolutePath
+  ): Future[Unit] = Future {
+    Files.walkFileTree(
+      sourceDirectory.toNIO,
+      new SimpleFileVisitor[Path] {
+        override def visitFile(
+            file: Path,
+            attrs: BasicFileAttributes
+        ): FileVisitResult = {
+          val path = AbsolutePath(file)
+          path.toLanguage match {
+            case Language.SCALA | Language.JAVA =>
+              index.addSourceFile(path, Some(sourceDirectory))
+            case _ =>
+          }
+          super.visitFile(file, attrs)
+        }
+        override def preVisitDirectory(
+            dir: Path,
+            attrs: BasicFileAttributes
+        ): FileVisitResult = {
+          val path = AbsolutePath(dir)
+          if (path.resolve("META-INF").isDirectory) {
+            FileVisitResult.SKIP_SUBTREE
+          } else if (dir.endsWith(".bloop") || dir.endsWith(".metals")) {
```

Comment:
This is actually no longer relevant because this method is only used for source directories now (which have no `.metals` or `.bloop`. Removed `preVisitDirectory`. Previously we ran this method at the workspace root, but that turned out to be a bad idea for many reasons.